agent/consul/util.go

```diff
@@ -273,89 +273,163 @@ func runtimeStats() map[string]string {
 	}
 }
 
-// ServersMeetMinimumVersion returns whether the given alive servers are at least on the
-// given Consul version
-func ServersMeetMinimumVersion(members []serf.Member, minVersion *version.Version) bool {
-	return ServersMeetRequirements(members, func(srv *metadata.Server) bool {
-		return srv.Status != serf.StatusAlive || !srv.Build.LessThan(minVersion)
-	})
+// checkServersProvider exists so that we can unit tests the requirements checking functions
+// without having to spin up a whole agent/server.
+type checkServersProvider interface {
+	CheckServers(datacenter string, fn func(*metadata.Server) bool)
+}
+
+// serverRequirementsFn should inspect the given metadata.Server struct
+// and return two booleans. The first indicates whether the given requirements
+// are met. The second indicates whether this server should be considered filtered.
+//
+// The reason for the two booleans is so that a requirement function could "filter"
+// out the left server members if we only want to consider things which are still
+// around or likely to come back (failed state).
+type serverRequirementFn func(*metadata.Server) (ok bool, filtered bool)
+
+type serversMeetRequirementsState struct {
+	// meetsRequirements is the callback to actual check for some specific requirement
+	meetsRequirements serverRequirementFn
+
+	// ok indicates whether all unfiltered servers meet the desired requirements
+	ok bool
+
+	// found is a boolean indicating that the meetsRequirement function did not filter
+	// at least one of the servers it was asked to process
```

Comment:
I found the wording of this a little confusing is this better?

```suggestion
	// found is a boolean indicating that the meetsRequirement function accepted at 
	// least one unfiltered server.
```